Components: Add support for TypeScript unit tests

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Following the exploratory work in #39341, this PR adds the necessary configuration in order to add unit tests written in TypeScript to the repo.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Having unit tests in TypeScript is a way to ensure that we type-check a component from the point of view of a consumer (and not just internally).

This PR will allow us to refactor to TypeScript unit tests in the `@wordpress/components` package, in the context of the wider package refactor (https://github.com/WordPress/gutenberg/issues/35744).

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- added a new TypeScript project specific for unit tests (`tsconfig.test.json`)
- added type defs for custom TypeScript matchers (`toMatchStyleDiffSnapshot`)
- edited `eslint` config to include TypeScript tests in the "development files" glob matcher
- refactored unit tests for the `<Text />` component to make sure that the above changes work for TypeScript unit tests

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

- Project build
- All unit tests pass
- Running `npm run build && npx tsc -p packages/components/tsconfig.json` doesn't output any errors to console.
- Open `packages/components/src/text/test/index.tsx` in an IDE supporting TypeScript:
     - make sure that type defs and autocomplete is working as expected.
     - introduce a TypeScript error, make sure it gets flagged in your IDE and when running `npx tsc -p packages/components/tsconfig.json`

ciampo committed:
Add test-specific `tsconfig`

ciampo committed:
Convert `Text` unit tests to TypeScript

ciampo committed:
Add custom Jest matcher types

ciampo committed:
Disable `import/no-extraneous-dependencies` for all TypeScript tests

ciampo committed:
Remove `@types/node` from `tsconfig.test.json`

ciampo commented:
Hey @sirreal — sorry for the ping again.

I opened this PR to introduce some of the changes discussed in #39341. 

But the fact that I chose to refactor to TypeScript another test (`packages/components/src/text/test/index.tsx`) seems to cause errors when running `npm run build:package-types `. Do you have any ideas on why this may be happening?

Maybe it's related to the fact that the new project for tests only includes files under `**/test/**` folders, and therefore types defined outside of these files can't be parsed?

dmsnell commented on typings/testing-library/index.d.ts:
```diff
@@ -0,0 +1,7 @@
+declare namespace jest {
+	interface Matchers< R > {
+		toMatchStyleDiffSnapshot( expected: Element ): CustomMatcherResult;
+	}
```
can we add an explanation for this declaration? it stands out in a number of ways to me that leave me confused:

 - Why does it exist at all?
 - What is the reason for the `R` type parameter?
 - Where doe we get the definition of `CustomMatcherResult`?

An explanatory comment could avoid a lot of confusion on those who are coming to this code and trying to understand it or modify it.

ciampo replied:
Hey @dmsnell , thank you for taking a look at the code!

This declaration exists to offer types for the `toMatchStyleDiffSnapshot` custom matcher (defined in [`test/unit/config/matchers/to-match-style-diff-snapshot.js`](https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/test/unit/config/matchers/to-match-style-diff-snapshot.js))

These are the definitions for the [`Matchers`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L690) and [`CustomMatcherResult`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L523-L526) interfaces.

After a few tries, I had landed on this format following [a few examples](https://github.com/testing-library/react-testing-library/issues/36#issuecomment-440442300) on GitHub. I've now pushed a [change](https://github.com/WordPress/gutenberg/pull/39436/commits/d58fce70c42eb0e548c2c60cc2c5806cbe8ea106) to align the syntax for this declaration with [the one used in the `snapshot-diff` package](https://github.com/jest-community/snapshot-diff/blob/master/index.d.ts#L12-L24).

sirreal replied:
It seems like `R` is the return type and probably comes from these returning `void|Promise<…>`. That's probably to allow both `expect().toBe()` and `expect().resolves.toBe()`.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d7d9c1d76dead09380e702d858b40432a29d8a48/types/jest/index.d.ts#L690

Did you consider putting these types in https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/test/unit/config/matchers/to-match-style-diff-snapshot.js? We might need to migrate that to TypeScript, but in my experience there's value in keeping the types close to what they're typing.

dmsnell replied:
thanks @ciampo! I appreciate the explanation but I found your comment here in GitHub clearer than the one in the code. Maybe that one in the code could use some language revision (read it out loud? have someone else read it and tell you what they think it means?)

otherwise, your comment here would make a great comment in the code. it doesn't really matter if I understand this while reviewing as much as it matters if the person who comes across this code in the future (could be yourself) gets the context for why it exists and what parts of it they can change without risking a problem.

```ts
/**
 * We use a custom matcher for the snapshot diff test which
 * takes into account project-specific behaviors and the
 * default matcher would invalidate those snapshots which
 * we consider valid.
 *
 * @see test/unit/config/matchers/to-match-style-diff-snapshot.js
 * @cite https://github.com/testing-library/react-testing-library/issues/36#issuecomment-440442300
 */
```

Using the `R` type parameter makes it obvious why it's there. Is `T` necessary?  What happens if we omit it? Will the interface declaration still merge? If not, we can add a sentence like this to the comment.

```ts
/**
 * We use a custom matcher for the snapshot diff test which
 * takes into account project-specific behaviors and the
 * default matcher would invalidate those snapshots which
 * we consider valid.
 *
 * The `T` type parameter is necessary to match the interface
 * declaration from the `jest` types even though it isn't used
 * in this section.
 *
 * @see test/unit/config/matchers/to-match-style-diff-snapshot.js
 * @cite https://github.com/testing-library/react-testing-library/issues/36#issuecomment-440442300
 */
```

ciampo replied:
Hey @sirreal and @dmsnell, thank you for your feedback here! I'm going to try and answer all of the points raised.

> It seems like R is the return type and probably comes from these returning void|Promise<…>. That's probably to allow both expect().toBe() and expect().resolves.toBe().

That is correct, I did some further exploration and came to the same conclusion

> Did you consider putting these types in [`test/unit/config/matchers/to-match-style-diff-snapshot.js`](https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/test/unit/config/matchers/to-match-style-diff-snapshot.js)? We might need to migrate that to TypeScript, but in my experience there's value in keeping the types close to what they're typing.

Things here get tricky. The signature of [the source `toMatchStyleDiffSnapshot` function](https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/test/unit/config/matchers/to-match-style-diff-snapshot.js) which gets passed to `expect.extend()` is different from the signature of the function effectively invoked in the unit test.

I have spent some time looking into this and found [this article](https://haspar.us/notes/adding-jest-custom-matchers-in-typescript#going-too-far-with-types) which illustrates this difference in the signatures, and even proposes a way to compute the "user-facing" function type given the source matcher function types.

The above technique would likely require us to rewrite all of the [jest config to TypeScript](https://jestjs.io/docs/getting-started#using-typescript), something that @sarayourfriend suggested in https://github.com/WordPress/gutenberg/pull/39341#issuecomment-1065400890. This wouldn't be a trivial task (for the best results, [`ts-jest`](https://github.com/kulshekhar/ts-jest) should be used over Babel), but **I could open a new PR to explore this direction, if we think it's a better overall solution ?**

>  I appreciate the explanation but I found your comment here in GitHub clearer than the one in the code.
> ...
> it doesn't really matter if I understand this while reviewing as much as it matters if the person who comes across this code in the future (could be yourself) gets the context for why it exists and what parts of it they can change without risking a problem.

Thank you for the feedback on this one. It is true that this comment needs to be clear for the next person that comes across this code, but they should also explain what the `toMatchStyleDiffSnapshot` matcher does when used in a test.

[I went ahead and updated the comment](https://github.com/WordPress/gutenberg/pull/39436/commits/fe79ab8cc77e21dff8fadb020a4487188d1a8465), taking inspiration from [the `toMatchSnapshot` comments](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d7d9c1d76dead09380e702d858b40432a29d8a48/types/jest/index.d.ts#L959-L962) and from the already existing docs that we have in [docs/contributors/code/testing-overview.md#best-practices](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/testing-overview.md#best-practices).

I also added the `@see` and `@cite` parts to the comment, as per your suggestion, although I didn't quite understand the part _"...and the default matcher would invalidate those snapshots which we consider valid."_. Could you explain further what you meant there?

I'd also be curious to hear @sarayourfriend 's opinion on what's the best description to add in the `toMatchStyleDiffSnapshot`'s docs.

> Using the R type parameter makes it obvious why it's there. Is T necessary? What happens if we omit it?

I removed the `T` parameter and it looks like everything still works as expected.

dmsnell replied:
On all points this sounds fair; thanks for taking the effort here to understand the types.

> Could you explain further what you meant there?

Your comment seems fine. Part of what makes mine awkward is that I understand the matcher type less than you do. I do have two quick nitpicks for the updated comment:

 - if we're using Markdown syntax, we might check if we can use the alternate link syntax since the inline `[title](URL)` format gets in the way of reading the code itself. if that works we should be able to follow the example below.
 - "associated to two different states" is probably smoother as "associated to _with_ different states" 

### alternate link syntax

```
/**
 * Similar to the [`toMatchDiffSnapshot` matcher], this custom matcher allows
 * to snapshot only the difference between the _styles_ associated to two
 * different states of a component.
 *
 * @see [to-match-style-diff-snapshot.js]
 * @see [Testing Overview docs]
 * @cite https://github.com/testing-library/react-testing-library/issues/36#issuecomment-440442300
 *
 * [`toMatchDiffSnapshot` matcher]: https://github.com/jest-community/snapshot-diff
 * [to-match-style-diff-snapshot.js]: https://github.com/WordPress/gutenberg/blob/trunk/test/unit/config/matchers/to-match-style-diff-snapshot.js
 * [Testing Overview docs]: https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/testing-overview.md#best-practices
 */
```

I'm not sure if our tools are capable of handling this syntax, so don't feel obliged to stick with it if it doesn't work out, but it would be nice to get rid of some of those long lines that are harder to parse when reading the source.

ciampo replied:
Thank you for the feedback — I had forgotten about the shorthand reference syntax for markdown links!

I applied the requested changes in https://github.com/WordPress/gutenberg/pull/39436/commits/2832c37688241c125d83ddbaf52686c23507258b and https://github.com/WordPress/gutenberg/pull/39436/commits/f52ec5fdaa7b6e8b2f609d70ee9bb16e6428fbe5 .

ciampo committed:
Update `toMatchStyleDiffSnapshot` type defs

ciampo commented:
Quick update:

> But the fact that I chose to refactor to TypeScript another test (`packages/components/src/text/test/index.tsx`) seems to cause errors when running `npm run build:package-types `. Do you have any ideas on why this may be happening?

I spent some time looking into these errors — it turns out that they are not _caused_ by this PR, they were only surfaced by it.

This can be observed by checking out `trunk`, running `npm run build:packages` and inspecting the files that were causing errors earlier in this PR:

- `packages/components/build-types/utils/config-values.d.ts`: on `trunk` this file actually contains a couple of circular dependency errors:

<details>
<summary>See screenshot</summary>

![Screenshot 2022-03-23 at 10 17 23](https://user-images.githubusercontent.com/1083581/159666030-090c5a6b-1e11-43c0-861b-adbbcddf5b9f.png)
</details>

- `packages/components/build-types/utils/colors-values.d.ts`: in this file, we can observe that effectively the type of the `COLORS` object doesn't include the `white` key.

In both cases, **I was finally able to solve the issue by using `Object.assign` instead of the `{ ...spread }` syntax.**

This should unlock further work on this PR (the CI checks should be able to build the project!). I will also look into addressing the feedback from this https://github.com/WordPress/gutenberg/pull/39436#discussion_r826459106

---

Nonetheless, **how can we make sure that the types that we generate when building our packages are error-free?** The error that was surfaced by this PR is currently on `trunk` but it's not being caught.

ciampo committed:
Change snapshot extension from `js` to `tsx`

ciampo committed:
Update toMatchStyleDiffSnapshot docs

ciampo committed:
Update test to use toMatchStyleDiffSnapshot

ciampo commented on typings/testing-library/index.d.ts:
```diff
@@ -1,11 +1,15 @@
+		 * @cite https://github.com/testing-library/react-testing-library/issues/36#issuecomment-440442300
 		 */
-		toMatchStyleDiffSnapshot( expected: Element ): R;
+		toMatchStyleDiffSnapshot( expected: Element | null ): R;
```
Just making a note here that we should review `toMatchStyleDiffSnapshot` to make sure that it behaves as expects if its arguments are `null`.

ciampo replied:
Done in https://github.com/WordPress/gutenberg/pull/39436/commits/748239563e8750cb74981dc9fee30160585c5ccf

ciampo committed:
Use alternative link syntax in the JSDoc comment

ciampo committed:
Improve comment wording

ciampo committed:
Allow null values to be passed to `toMatchStyleDiffSnapshot`

ciampo commented:
Update:
- I addressed the feedback regarding the type declaration for `toMatchStyleDiffSnapshot` 
- I moved the changes described in [this previous comment](https://github.com/WordPress/gutenberg/pull/39436#issuecomment-1076150261) to a separate PR #39932. That PR has been recently merged, and I rebased this PR to include those changes.


I believe that this PR is ready for a final round of review if we think that this approach is good enough to move forward.

<details>

<summary>Here's a summary of the other approach that was discussed (click to explan)</summary>

To recap, the main alternative approach that was discussed was to rewrite all of the [jest config to TypeScript](https://jestjs.io/docs/getting-started#using-typescript) (as also suggested previously in https://github.com/WordPress/gutenberg/pull/39341#issuecomment-1065400890).

This wouldn't be a trivial task (for the best results, [`ts-jest`](https://github.com/kulshekhar/ts-jest) should be used over Babel), but it would likely enable us to use most "first-party" types without having to explicitly get from `definitely-typed` and/or defining them in `tsconfig.test.json`.

It would also allow us to rewrite the custom `toMatchStyleDiffSnapshot` matcher in TypeScript and follow a similar approach to what is described in [this article](https://haspar.us/notes/adding-jest-custom-matchers-in-typescript#going-too-far-with-types) to define its type.
</details>

dmsnell commented:
Thanks for engaging with my feedback. I don't feel like there's anything else I can contribute review-wise because I don't understand the implications fully of the changes to the main `tsconfig.json` config. It seems like it'll be worth a try if nobody else has any objections.

sirreal commented on typings/testing-library/index.d.ts:
```diff
@@ -0,0 +1,20 @@
+/// <reference types="jest"/>
+
```
I don't believe this reference is necessary. The namespace and interface will be merged as they come into scope.

```suggestion
```

sirreal commented on typings/testing-library/index.d.ts:
```diff
@@ -0,0 +1,20 @@
+/// <reference types="jest"/>
+
+declare namespace jest {
```
This is from Gutenberg's test utils, correct? Naming this file `testing-library` seems confusing to me. I'd change the name to `gutenberg-test-env` or similar to make it clear that this is Gutenberg-specific and not a library. That also aligns with the runtime env `gutenberg-env`.

sirreal committed:
Remove jest types reference

sirreal committed:
Rename testing-library declarations to gutenberg-test-env

mirka reviewed (approved):
It took some futzing around for VS Code to acknowledge the `toMatchStyleDiffSnapshot` typing for me, but once it did, everything is working as expected 💥

I'm very happy about this, thanks for working hard on it!

sirreal reviewed (commented):
I'll push changes with these suggestions.

sirreal reviewed (changes requested):
I discovered a wrinkle while testing this out now. First, apply this patch:

```diff
diff --git a/tsconfig.test.json b/tsconfig.test.json
index 85ff605e1d..433b1274d4 100644
--- a/tsconfig.test.json
+++ b/tsconfig.test.json
@@ -1,6 +1,8 @@
 {
 	"extends": "./tsconfig.base.json",
 	"compilerOptions": {
+		"allowJs": false,
+		"checkJs": false,
 		"noEmit": true,
 		"emitDeclarationOnly": false,
 		"types": [
```

This seems straightforward. We only want to typecheck TypeScript test files so we change this from the settings in the extended config. Now let's compile the test project: `node_modules/.bin/tsc --build tsconfig.test.json`:

```
packages/components/src/text/test/index.tsx:10:24 - error TS7016: Could not find a declaration file for module '../../utils'. 'packages/components/src/utils/index.js' implicitly has an 'any' type.

10 import { COLORS } from '../../utils';
                          ~~~~~~~~~~~~~

packages/components/src/text/test/index.tsx:11:22 - error TS7016: Could not find a declaration file for module '../'. 'packages/components/src/text/index.js' implicitly has an 'any' type.

11 import { Text } from '../';
                        ~~~~~


Found 2 errors.
```

Oops 🙂 

What's happening?

We expected to compile all the application code, that's why we added references to every single project. Then, the tests should discover and use the declarations that were generated.

What happened instead? The short version is that we're accessing files from tests like `import { Text } from '../`. No declaration file exists for the file, so `tsc` tries to build it. Even though we configured our project to only include files immediately under `test` directories, the compiler finds that import so it becomes part of the project and we try to build it.

As-is, this PR compiles _all the projects_ in a first pass, then recompiles just a couple files (including application code) of the files that are part of the test dependency graph. That isn't what we intended at all!

I was hoping that this problem was limited to `@wordpress/components` because it's missing the `types` field in package.json that points to the `build-types` directory. Unfortunately that does not seem to be the case 😞 Because we're importing relative modules (`../…`) from inside the same package, TypeScript doesn't have a chance to resolve these files from the package's type declarations 😬 

I'm not sure what the best way forward is right now, I need to think about it a bit and do some testing. I'd be happy to chat about the problem and discuss possible solutions.

ciampo commented:
Does it mean that, basically, a test written in TypeScript should have all of its local dependencies also written in TypeScript? Would adding `// @ts-ignore` at the start of these non-TS files help?

sirreal commented:
> Does it mean that, basically, a test written in TypeScript should have all of its local dependencies also written in TypeScript?

No. The problem I'm highlighting is that:

- We're compiling non-test source files twice (once for the project, and again as part of the test project).
- We're not compiling the non-test source files according to their projects, but as part of the test project.

We expect to import `.js` files, but we'd like to use the declarations that were generated for them.

> Would adding `// @ts-ignore` at the start of these non-TS files help?

No. It's possible it could make errors go away, but I think we'd just lose type information for those files (imports from those modules would become `any`) which is not desirable.

ciampo commented:
Thank you for the clarification, @sirreal !  I'll go ahead and ask another question (I feel like I'm kind of rubber-ducking at this point): would the initial approach that I took in the previous PR (https://github.com/WordPress/gutenberg/pull/39341/commits/d471426dbe8a85569492fb3934aacc02cc2ec6f0) be an idea? Basically using the existing package's `tsconfig.json` and including the tests file to it?

sirreal commented:
> using the existing package's tsconfig.json and including the tests file to it?

I think that makes sense at this point. The main drawback is that we pollute the application code with test types (jest globals) but in practice that doesn't usually cause problems.

ciampo commented:
Alright, I'll proceed an implement this idea in the next days — of course feel free to propose other approaches if they come to mind!

ciampo committed:
Change approach from a dedicated test project, to re-using the components' existing project

ciampo commented:
Rebased, and [pushed changes](https://github.com/WordPress/gutenberg/pull/39436/commits/e0c3506768e47ad6ae4a5ad5898783136d233199) to move away from a dedicated `tsconfig.test.json` and re-use the existing components package's `tsconfig.json`.

@sirreal and @mirka , could you have another look?

sirreal reviewed (approved):
This looks good to me, thanks @ciampo!
